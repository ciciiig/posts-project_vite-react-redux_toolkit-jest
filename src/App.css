.app-container {
    box-sizing: border-box;
    border: 1px dashed black;
    height: 90vh;
    width: 90vw;
    margin: 30px auto;
    padding: 30px;
}

.search-input {
    border: 2px solid black;
    height: 40px;
    padding: 15px;
}

.posts_container {
    display: flex;
    justify-content: start;
    align-content: flex-start;
    flex-wrap: wrap;
    border: 2px solid black;
    height: 80%;
    margin: 10px 0;
    padding: 15px;
}

.post {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    border: 1px solid black;
    max-width: 120px;
    height: 120px;
    margin: 5px;
    padding: 5px;
    width: 100%;
}

.post:hover {
    transform: scale(1.02);
    cursor: pointer;
}

.post-body {
    /* line-clamp needs all 4 lines bellow to show only 3 lines */
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
}

/* .post-navigation-container {
    text-align: center;
}

.page-number {
    margin: 10px;
    pointer-events: none;
}

.arrow {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    font: inherit;
    cursor: pointer;
    outline: inherit;
}

.arrow:disabled {
    color: rgba(150, 150, 150, 0.5);
    pointer-events: none;
} */

.modal-back {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: rgba(0, 0, 0, .75);
    z-index: 10000;
}

.modal-window {
    max-height: 350px;
    width: 450px;
    background-color: white;
    border: 2px solid black;
    padding: 16px;
}

.modal-window-header {
    display: flex;
    justify-content: flex-end;
}

.modal-window-acton-buttons-container {
    margin: 1em 0;
    display: flex;
    justify-content: center;
}

.alert-container {
    position: absolute;
    top: 15%;  
    left: 50%; 
    transform: translate(-50%, -50%);
    background-color: rgb(228, 228, 228);
}

.alert-message_updating {
    border: 2.5px dashed blue;
    padding: 10px;
}

.alert-message_error {
    border: 2.5px solid red;
    padding: 10px;
}