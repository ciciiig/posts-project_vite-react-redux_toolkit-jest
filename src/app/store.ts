import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit"
import counterReducer from "../features/counter/counterSlice"
import paginationReducer from "../features/pagination/paginationSlice"
import postsReducer from "../features/posts/postsSlice"

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    pagination: paginationReducer,
    posts: postsReducer,
  },
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
