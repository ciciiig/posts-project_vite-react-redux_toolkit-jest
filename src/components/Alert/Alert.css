.alert-container {
    position: absolute;
    top: 15%;  
    left: 50%; 
    transform: translate(-50%, -50%);
    background-color: rgb(228, 228, 228);
}

.alert-message_updating {
    border: 2.5px dashed blue;
    padding: 10px;
}

.alert-message_error {
    border: 2.5px solid red;
    padding: 10px;
}