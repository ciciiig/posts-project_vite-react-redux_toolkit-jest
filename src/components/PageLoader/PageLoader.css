.alertSkeleton-container {
    background-color: rgba(92, 60, 60, 0.88);
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    justify-content: center;
    align-items: center;
    color: white;
    font-size: 46px;
}
