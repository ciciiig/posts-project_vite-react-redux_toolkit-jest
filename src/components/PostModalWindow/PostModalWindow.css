.modal-back {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: rgba(0, 0, 0, .75);
    z-index: 10000;
}

.modal-window {
    max-height: 350px;
    width: 450px;
    background-color: white;
    border: 2px solid black;
    padding: 16px;
}

.modal-window-header {
    display: flex;
    justify-content: flex-end;
}

.modal-window-acton-buttons-container {
    margin: 1em 0;
    display: flex;
    justify-content: center;
}