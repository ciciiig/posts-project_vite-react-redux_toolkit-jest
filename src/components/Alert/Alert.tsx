import "./Alert.css"
import { FC, ReactNode, useEffect, useState } from "react"
import { useAppSelector } from "../../redux/hooks"
import { selectPosts } from "../../redux/posts"

export const Alert: FC = () => {
  const { status, error } = useAppSelector(selectPosts)
  const [alertState, setAlertState] = useState<{
    isShow: boolean
    alertContent: ReactNode
  }>({
    isShow: false,
    alertContent: null,
  })

  // For close Alert
  useEffect(() => {
    let timerId: number

    if (alertState.isShow) {
      timerId = setTimeout(() => {
        setAlertState({ isShow: false, alertContent: null })
        clearTimeout(timerId)
      }, 3000)
    }

    return () => {
      timerId && clearTimeout(timerId)
    }
  }, [alertState.isShow])

  // For set open Alert with correct content
  useEffect(() => {
    if (status === "loading") {
      setAlertState({
        isShow: true,
        alertContent: <div className="alert-message_updating">Loading...</div>,
      })
    }

    if (status === "failed") {
      setAlertState({
        isShow: true,
        alertContent: (
          <div className="alert-message_error">{error ?? "Some error"}</div>
        ),
      })
    }
  }, [status])

  return alertState.isShow ? (
    <div className="alert-container" id="alert-container">
      {alertState.alertContent}
    </div>
  ) : null
}
