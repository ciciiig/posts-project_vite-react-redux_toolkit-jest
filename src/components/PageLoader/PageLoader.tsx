import "./PageLoader.css"
import { FC } from "react"
import { useAppSelector } from "../../redux/hooks"
import { selectPosts } from "../../redux/posts"

export const PageLoader: FC = () => {
  const postState = useAppSelector(selectPosts)
  let skeletonAlertContent = null

  if (postState.skeletonStatus === "loading") {
    skeletonAlertContent = (
      <div className="alertSkeleton-container">Loading...</div>
    )
  }

  if (postState.skeletonStatus === "failed") {
    skeletonAlertContent = (
      <div className="alertSkeleton-container">
        <div>{postState.error ?? "Some error"}</div>
      </div>
    )
  }

  if (postState.skeletonStatus === "idle") {
    skeletonAlertContent = null
  }

  return skeletonAlertContent
}
