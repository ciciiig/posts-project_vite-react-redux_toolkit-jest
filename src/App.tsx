import "./App.css"
import { Input } from "./components/Input/Input"
import { Pagination } from "./components/Pagination/Pagination"
import { PostsList } from "./components/PostsList/PostsList"
import { PostModalWindow } from "./components/PostModalWindow/PostModalWindow"
import { useAppSelector } from "./redux/hooks"
import { selectPostModal } from "./redux/postModal"
import { Alert } from "./components/Alert/Alert"
import { PageLoader } from "./components/PageLoader/PageLoader"

function App() {
  const { isOpen: isPostModalWindowOpen } = useAppSelector(selectPostModal)

  return (
    <div className="app-container" id="app-container">
      {isPostModalWindowOpen && <PostModalWindow />}

      <>
        <Input />
        <PostsList />
        <Pagination />
      </>

      <PageLoader />
      <Alert />
    </div>
  )
}

export default App
