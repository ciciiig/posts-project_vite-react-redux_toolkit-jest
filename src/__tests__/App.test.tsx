import App from "../App"
import { PostModalWindow } from "../components/PostModalWindow/PostModalWindow"
import { useAppSelector } from "../redux/hooks"
import { Alert } from "../components/Alert/Alert"
import { PageLoader } from "../components/PageLoader/PageLoader"
import { Posts } from "../components/Posts/Posts"

import { render } from "@testing-library/react"
import { RootState } from "../redux/store"
import { clone } from "ramda"

jest.mock("../components/PostModalWindow/PostModalWindow", () => ({
  ...jest.requireActual("../components/PostModalWindow/PostModalWindow"),
  PostModalWindow: jest.fn(),
}))
jest.mock("../redux/hooks", () => ({
  ...jest.requireActual("../redux/hooks"),
  useAppSelector: jest.fn(),
}))
jest.mock("../components/Alert/Alert", () => ({
  ...jest.requireActual("../components/Alert/Alert"),
  Alert: jest.fn(),
}))
jest.mock("../components/PageLoader/PageLoader", () => ({
  ...jest.requireActual("../components/PageLoader/PageLoader"),
  PageLoader: jest.fn(),
}))
jest.mock("../components/Posts/Posts", () => ({
  ...jest.requireActual("../components/Posts/Posts"),
  Posts: jest.fn(),
}))

describe("Test <App />", () => {
  const initialState = {
    postModal: {
      isOpen: true,
    },
  } as RootState
  let currentState: RootState

  beforeEach(() => {
    jest.clearAllMocks()

    currentState = clone(initialState)

    jest.mocked(useAppSelector).mockImplementation((fn) => fn(currentState))
    jest
      .mocked(PostModalWindow)
      .mockImplementation(() => <div>mockedPostModalWindow</div>)
    jest.mocked(Posts).mockImplementation(() => <div>mockedPosts</div>)
    jest
      .mocked(PageLoader)
      .mockImplementation(() => <div>mockedPageLoader</div>)
    jest.mocked(Alert).mockImplementation(() => <div>mockedAlert</div>)
  })

  it("show modal window if isOpen is true", () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })

  it("Don't show modal window if isOpen is false", () => {
    currentState.postModal.isOpen = false

    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })
})
