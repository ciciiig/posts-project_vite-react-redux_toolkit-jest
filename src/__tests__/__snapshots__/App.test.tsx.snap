// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <App /> Don't show modal window if isOpen is false 1`] = `
<div>
  <div
    class="app-container"
    id="app-container"
  >
    <div>
      mockedPosts
    </div>
    <div>
      mockedPageLoader
    </div>
    <div>
      mockedAlert
    </div>
  </div>
</div>
`;

exports[`Test <App /> show modal window if isOpen is true 1`] = `
<div>
  <div
    class="app-container"
    id="app-container"
  >
    <div>
      mockedPostModalWindow
    </div>
    <div>
      mockedPosts
    </div>
    <div>
      mockedPageLoader
    </div>
    <div>
      mockedAlert
    </div>
  </div>
</div>
`;
